OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs

```diff
@@ -66,55 +65,43 @@ public void Tick(Actor self)
 				return;
 
 			VisibilityHash = 0;
-			foreach (var p in self.World.Players)
-			{
-				// We are doing the following LINQ manually to avoid allocating an extra delegate since this is a hot path.
-				// var isVisible = footprint.Any(p.Shroud.IsVisibleTest(footprintRegion));
-				var isVisibleTest = p.Shroud.IsVisibleTest(footprintRegion);
-				var isVisible = false;
-				foreach (var uv in footprint)
-					if (isVisibleTest(uv))
-					{
-						isVisible = true;
-						break;
-					}
-
-				visible[p] = isVisible;
-				if (isVisible)
-					VisibilityHash += p.ClientIndex;
-			}
-
-			if (!initialized)
+			foreach (var player in self.World.Players)
 			{
-				foreach (var p in self.World.Players)
+				bool isVisible;
+				FrozenActor frozenActor;
+				if (!initialized)
 				{
-					visible[p] |= startsRevealed;
-					p.PlayerActor.Trait<FrozenActorLayer>().Add(frozen[p] = new FrozenActor(self, footprint, footprintRegion));
+					frozen[player] = frozenActor = new FrozenActor(self, footprint, footprintRegion, player.Shroud);
+					player.PlayerActor.Trait<FrozenActorLayer>().Add(frozenActor);
```

Comment:
You could cache the trait lookups for all `self.World.Players` in the constructor to save even more resources.